File: java/com/tme_inc/pwv/TagEvent.java
package com.tme_inc.pwv;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;
import android.view.Menu;
import android.view.MenuItem;
import android.view.View;
import android.widget.ArrayAdapter;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.Spinner;

import java.util.Arrays;
import java.util.Calendar;

public class TagEvent extends Activity {

    private PWProtocol mPwProtocol ;
    private int VriItemSize=0 ;

    // DVR Date/Time in BCD
    private long DvrDate = 0 ;
    private long DvrTime = 0 ;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        long timestamp = getIntent().getLongExtra("DvrTime", 0L);

        Calendar calendar = Calendar.getInstance();
        if( timestamp>0L ) {
            calendar.clear();
            calendar.setTimeInMillis(timestamp);
        }
        DvrDate = calendar.get(Calendar.YEAR) * 10000 +
                (calendar.get(Calendar.MONTH) - Calendar.JANUARY + 1) * 100 +
                calendar.get(Calendar.DAY_OF_MONTH) ;
        DvrTime = calendar.get(Calendar.HOUR_OF_DAY) * 10000 +
                calendar.get(Calendar.MINUTE) * 100 +
                calendar.get(Calendar.SECOND) ;


        setContentView(R.layout.activity_tag_event);

        // init incident list
        Spinner spinner ;
        spinner = (Spinner)findViewById(R.id.tag_incident);
        if( spinner!=null ) {
            ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this,
                    R.array.tag_incident_list, android.R.layout.simple_spinner_item);
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            spinner.setAdapter(adapter);
        }

        // init priority list
        spinner = (Spinner)findViewById(R.id.tag_priority);
        if( spinner!=null ) {
            ArrayAdapter<CharSequence> adapter = ArrayAdapter.createFromResource(this,
                    R.array.tag_priority_list, android.R.layout.simple_spinner_item);
            adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
            spinner.setAdapter(adapter);
        }

        // init PW event processor
        //
        mPwProtocol = new PWProtocol();
        mPwProtocol.GetVri( new PWProtocol.PWListener() {
            @Override
            public void onPWEvent(Bundle result) {
                onTagPWEvent(result);
            }
        });

        Button button = (Button)findViewById(R.id.button_tag) ;
        button.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onTagEventClick();
            }
        });

        button= (Button)findViewById(R.id.button_cancel) ;
        button.setOnClickListener(new View.OnClickListener() {
                                      @Override
                                      public void onClick(View v) {
                                          onCancelClick();
                                      }
                                  });

    }


    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        // Inflate the menu; this adds items to the action bar if it is present.
        getMenuInflater().inflate(R.menu.menu_tag_event, menu);
        return true;
    }

    @Override
    public boolean onOptionsItemSelected(MenuItem item) {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        int id = item.getItemId();

        //noinspection SimplifiableIfStatement
        if (id == R.id.action_settings) {
            return true;
        }

        return super.onOptionsItemSelected(item);
    }

    protected void onTagPWEvent(Bundle result) {
        if ( result!=null ) {
            VriItemSize = result.getInt("VriItemSize", 0);
            int VriSize = result.getInt("VriListSize", 0);
            byte [] VriList = result.getByteArray("VriList");
            if (VriSize>0 && VriItemSize>0 && VriList!=null ) {
                int offset = 0 ;

                long searchdatetime = (DvrDate - 20000000) * 10000 + DvrTime/100  ;
                //search for current index
                long vr_p =0 ;
                int vr_pi = 0 ;
                long vr_n = 0 ;
                int vr_ni = 0;
                int o ;

                // Search matching vri
                for( o=0; o<(VriSize*VriItemSize); o+=VriItemSize ) {
                    long vridatetime = 0 ;

                    //VRI
                    String vri =  new String(VriList, o, 64).split("\0")[0].trim();
                    // Split vri date/time
                    String [] vriarray = vri.split("-");
                    if( vriarray.length>1 ) {
                        try{
                            vridatetime = Long.parseLong(vriarray[1]) ;
                        }catch(NumberFormatException e){
                            vridatetime = 0;
                        }
                        catch( Exception e ) {
                            vridatetime = 0 ;
                        }
                    }

                    if( vridatetime<=searchdatetime ) {
                        if( vr_p == 0 ) {
                            vr_p = vridatetime ;
                            vr_pi = o ;
                        }
                        else if( vridatetime > vr_p ) {
                            vr_p = vridatetime ;
                            vr_pi = o ;
                        }
                    }
                    else {
                        if( vr_n == 0 ) {
                            vr_n = vridatetime ;
                            vr_ni = o ;
                        }
                        else if( vridatetime < vr_n ) {
                            vr_n = vridatetime ;
                            vr_ni = o ;
                        }
                    }
                }

                if( vr_p !=0 ) {
                    o=vr_pi ;
                } else if( vr_n!=0 ) {
                    o=vr_ni ;
                }
                else {
                    o= 0 ;
                }


                offset = o ;

                String OfficerId = "";
                String str =  new String(VriList, offset, 64).split("\0")[0].trim();
                EditText eVri = (EditText)findViewById(R.id.tag_vri);
                if( eVri!=null ) {
                    eVri.setText(str);
                }
                // splite officer ID from vri
                String [] strarray = str.split("-");
                if( strarray.length>2 ) {
                    OfficerId = strarray[strarray.length-1];
                }
                offset+=64 ;

                // incident classification
                str =  new String(VriList, offset, 32).split("\0")[0].trim();
                Spinner sIncident = (Spinner)findViewById(R.id.tag_incident);
                if( sIncident!=null ) {
                    ArrayAdapter<CharSequence> adapter = (ArrayAdapter<CharSequence>)sIncident.getAdapter();
                    int pos = adapter.getPosition(str);
                    if(pos<0) {
//                                    adapter.add(str);
//                                    pos = adapter.getPosition(str);
                    }
                    sIncident.setSelection(pos);
                }
                offset+=32 ;

                // case number
                str =  new String(VriList, offset, 64).split("\0")[0].trim();
                EditText eCase = (EditText)findViewById(R.id.tag_casenumber);
                if( eCase!=null ) {
                    eCase.setText(str);
                }
                offset+=64 ;

                // Priority
                str =  new String(VriList, offset, 20).split("\0")[0].trim();
                Spinner sPrio = (Spinner)findViewById(R.id.tag_priority);
                if( sPrio!=null ) {
                    ArrayAdapter<CharSequence> adapter = (ArrayAdapter<CharSequence>)sPrio.getAdapter();
                    int pos = adapter.getPosition(str);
                    if(pos<0) {
//                                    adapter.add(str);
//                                    pos = adapter.getPosition(str);
                    }
                    sPrio.setSelection(pos);
                }
                offset+=20 ;

                // Officer ID
                EditText eOfficerId = (EditText)findViewById(R.id.tag_officerid);
                if( eOfficerId!=null ) {
                    eOfficerId.setText(OfficerId);
                }
                offset+=32 ;    // skip officer ID

                // Notes
                str =  new String(VriList, offset, 255).split("\0")[0].trim();
                EditText eNotes = (EditText)findViewById(R.id.tag_notes);
                if( eNotes!=null ) {
                    eNotes.setText(str);
                }

            }

        }
        return;
    }

    protected  void onTagEventClick()
    {
        if (VriItemSize > 0) {
            byte[] vri = new byte[VriItemSize];
            Arrays.fill(vri, (byte) ' ');

            String str;
            int len;
            int offset = 0;

            // Vri
            EditText eVri = (EditText) findViewById(R.id.tag_vri);
            if (eVri != null) {
                byte[] bVri = eVri.getText().toString().getBytes();
                len = bVri.length;
                if (len > 64) len = 64;
                System.arraycopy(bVri, 0, vri, offset, len);
            }
            offset += 64;

            // incident classification
            Spinner sIncident = (Spinner) findViewById(R.id.tag_incident);
            if (sIncident != null) {
                str = (String) sIncident.getSelectedItem();
                byte[] bIncident = str.getBytes();
                len = bIncident.length;
                if (len > 32) len = 32;
                System.arraycopy(bIncident, 0, vri, offset, len);
            }
            offset += 32;

            // case number
            EditText eCase = (EditText) findViewById(R.id.tag_casenumber);
            if (eCase != null) {
                byte[] bCase = eCase.getText().toString().getBytes();
                len = bCase.length;
                if (len > 64) len = 64;
                System.arraycopy(bCase, 0, vri, offset, len);
            }
            offset += 64;

            // Priority
            Spinner sPrio = (Spinner) findViewById(R.id.tag_priority);
            if (sPrio != null) {
                str = (String) sPrio.getSelectedItem();
                byte[] bPrio = str.getBytes();
                len = bPrio.length;
                if (len > 20) len = 20;
                System.arraycopy(bPrio, 0, vri, offset, len);
            }
            offset += 20;

            // Officer ID
            offset += 32;    // skip officer ID

            // Notes
            EditText eNotes = (EditText) findViewById(R.id.tag_notes);
            if (eNotes != null) {
                byte[] bNotes = eNotes.getText().toString().getBytes();
                len = bNotes.length;
                int maxlen = VriItemSize - offset;
                if (len > maxlen) len = maxlen;
                System.arraycopy(bNotes, 0, vri, offset, len);
            }

            // set VRI
            if (mPwProtocol != null) {
                mPwProtocol.SetVri(vri, null);
            }

        }
        finish();
    }

    protected void onCancelClick() {
        finish();
    }

}

File: java/com/tme_inc/pwv/TagEvent.kt
package com.tme_inc.pwv

import android.app.Activity
import android.os.Bundle
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.ArrayAdapter
import android.widget.Button
import android.widget.EditText
import android.widget.Spinner
import kotlinx.android.synthetic.main.activity_tag_event.*

import java.util.Arrays
import java.util.Calendar
import kotlin.math.max
import kotlin.math.min

class TagEvent : Activity() {

    private var mPwProtocol: PWProtocol? = null
    private var VriItemSize = 0

    // DVR Date/Time in BCD
    private var DvrDate: Long = 0
    private var DvrTime: Long = 0


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val timestamp = intent.getLongExtra("DvrTime", 0L)

        val calendar = Calendar.getInstance()
        if (timestamp > 0L) {
            calendar.clear()
            calendar.timeInMillis = timestamp
        }
        DvrDate = (calendar.get(Calendar.YEAR) * 10000 +
                (calendar.get(Calendar.MONTH) - Calendar.JANUARY + 1) * 100 +
                calendar.get(Calendar.DAY_OF_MONTH)).toLong()
        DvrTime = (calendar.get(Calendar.HOUR_OF_DAY) * 10000 +
                calendar.get(Calendar.MINUTE) * 100 +
                calendar.get(Calendar.SECOND)).toLong()


        setContentView(R.layout.activity_tag_event)

        // init incident list
        var adapter = ArrayAdapter.createFromResource(
            this,
            R.array.tag_incident_list, android.R.layout.simple_spinner_item
        )
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        tag_incident.adapter = adapter

        // init priority list
        adapter = ArrayAdapter.createFromResource(
            this,
            R.array.tag_priority_list, android.R.layout.simple_spinner_item
        )
        adapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        tag_priority.adapter = adapter

        // init PW event processor
        //
        mPwProtocol = PWProtocol()
        mPwProtocol!!.getVri {
            onTagPWEvent(it)
        }

        button_tag.setOnClickListener {
            onTagEventClick()
        }

        button_cancel.setOnClickListener {
            finish()
        }

    }


    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        // Inflate the menu; this adds items to the action bar if it is present.
        menuInflater.inflate(R.menu.menu_tag_event, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        // Handle action bar item clicks here. The action bar will
        // automatically handle clicks on the Home/Up button, so long
        // as you specify a parent activity in AndroidManifest.xml.
        return if (item.itemId == R.id.action_settings) {
            true
        } else super.onOptionsItemSelected(item)

    }

    private fun getVriTime( vri : String ) : Long {
        //trim VRI
        val tVri = vri.split("\u0000")[0].trim()
        val vriarray = tVri.split("-")
        if (vriarray.size > 1) {
            var vriTime: Long = 0
            try {
                vriTime = vriarray[vriarray.size - 1].toLong()        // this one could be police ID
                if (vriTime > 1000000000 && vriTime < 5000000000) {
                    return vriTime
                }
            } catch (e: NumberFormatException) {
                vriTime = 0
            } catch (e: Exception) {
                vriTime = 0
            }

            try {
                vriTime = vriarray[vriarray.size - 2].toLong()
                if (vriTime > 1000000000 && vriTime < 5000000000) {
                    return vriTime
                }
            } catch (e: NumberFormatException) {
                vriTime = 0
            } catch (e: Exception) {
                vriTime = 0
            }
            return vriTime
        }
        return 0L
    }

    protected fun onTagPWEvent(result: Bundle?) {
        if (result != null) {
            VriItemSize = result.getInt("VriItemSize", 0)
            val VriSize = result.getInt("VriListSize", 0)
            val VriList = result.getByteArray("VriList")
            if (VriSize > 0 && VriItemSize > 0 && VriList != null) {

                val searchdatetime = (DvrDate - 20000000) * 10000 + DvrTime / 100
                //search for current index
                var vr_p: Long = 0
                var vr_pi = 0
                var vr_n: Long = 0
                var vr_ni = 0

                // extension of vri
                fun ByteArray.getVriItem(offset: Int, len: Int ) : String {
                    return String(
                        this,
                        offset,
                        len
                    ).split("\u0000")[0].trim()
                }

                // Search matching vri
                var offset = 0
                while (offset < VriSize * VriItemSize) {

                    // VRI time
                    var vridatetime: Long = getVriTime( String(
                        VriList,
                        offset,
                        64
                    ))

                    if (vridatetime <= searchdatetime) {
                        if (vr_p == 0L) {
                            vr_p = vridatetime
                            vr_pi = offset
                        } else if (vridatetime > vr_p) {
                            vr_p = vridatetime
                            vr_pi = offset
                        }
                    } else {
                        if (vr_n == 0L) {
                            vr_n = vridatetime
                            vr_ni = offset
                        } else if (vridatetime < vr_n) {
                            vr_n = vridatetime
                            vr_ni = offset
                        }
                    }
                    offset += VriItemSize
                }

                offset = if (vr_p != 0L) {
                    vr_pi
                } else if (vr_n != 0L) {
                    vr_ni
                } else {
                    0
                }

                var OfficerId = ""

                var str = VriList.getVriItem(offset, 64)
                tag_vri.setText(str)
                // splite officer ID from vri
                val strarray = str.split("-")
                if (strarray.size > 2) {
                    OfficerId = strarray[strarray.size - 1]
                }
                offset += 64

                var apos = 0

                // incident classification
                str = VriList.getVriItem(offset, 32)
                val incident_adapter = tag_incident.adapter as ArrayAdapter<CharSequence>
                apos = incident_adapter.getPosition(str)
                if (apos < 0) {
                    //                                    adapter.add(str);
                    //                                    pos = adapter.getPosition(str);
                    apos = 0
                }
                tag_incident.setSelection(apos)
                offset += 32

                // case number
                tag_casenumber.setText(VriList.getVriItem(offset, 64))
                offset += 64

                // Priority
                str = VriList.getVriItem(offset, 20)
                val prio_adapter = tag_priority.adapter as ArrayAdapter<CharSequence>
                apos = prio_adapter.getPosition(str)
                if (apos < 0) {
                    //                                    adapter.add(str);
                    //                                    pos = adapter.getPosition(str);
                    apos = 0
                }
                tag_priority.setSelection(apos)
                offset += 20

                // Officer ID
                tag_officerid.setText(OfficerId)
                offset += 32    // skip officer ID

                // Notes
                tag_notes.setText(VriList.getVriItem(offset, 255))

            }

        }
        return
    }

    private fun onTagEventClick() {
        if (VriItemSize > 0) {
            val vri = ByteArray(VriItemSize) {
                ' '.toByte()
            }

            fun ByteArray.setVriItem( sData : String, offset: Int, len: Int) : Int {
                val bData = sData.toByteArray()
                val blen = min( bData.size, len )
                System.arraycopy(bData, 0, this, offset, blen)
                return len
            }

            var offset = 0

            // Vri
            offset += vri.setVriItem(tag_vri.text.toString(), offset, 64)

            // incident classification
            offset += vri.setVriItem(tag_incident.selectedItem as String, offset, 32)

            // case number
            offset += vri.setVriItem(tag_casenumber.text.toString(), offset, 64)

            // Priority
            offset += vri.setVriItem(tag_priority.selectedItem.toString(), offset, 20)

            // Officer ID
            offset += 32    // skip officer ID

            // Notes
            vri.setVriItem(tag_notes.text.toString(), offset, VriItemSize - offset)

            // set VRI
            if (mPwProtocol != null) {
                mPwProtocol!!.setVri(vri)
            }

        }
        finish()
    }

}
